test/sslapitest.c

```diff
@@ -1873,36 +1878,142 @@ static int ocsp_server_cb(SSL *s, void *arg)
         return SSL_TLSEXT_ERR_ALERT_FATAL;
     }
 
-    if (!TEST_ptr(copy = OPENSSL_memdup(orespder, sizeof(orespder))))
+    dummy_ocsp_resp = OCSP_response_create(OCSP_RESPONSE_STATUS_SUCCESSFUL, NULL);
+
+    if (!TEST_ptr(dummy_ocsp_resp)) {
         return SSL_TLSEXT_ERR_ALERT_FATAL;
+    }
 
-    if (!TEST_true(SSL_set_tlsext_status_ocsp_resp(s, copy,
-                                                   sizeof(orespder)))) {
-        OPENSSL_free(copy);
+    resplen = i2d_OCSP_RESPONSE(dummy_ocsp_resp, &respder);
+
+    OCSP_RESPONSE_free(dummy_ocsp_resp);
+
+    if (!TEST_true(SSL_set_tlsext_status_ocsp_resp(s, respder, resplen))) {
+        OPENSSL_free(respder);
         return SSL_TLSEXT_ERR_ALERT_FATAL;
     }
+
+    OPENSSL_free(respder);
+
```

Comment:
SSL_set_tlsext_status_ocsp_resp() API should consume respder in order to preserve API old behavior.

Reply:
Hi Martin, just noted that in old API  : SSL_set_tlsext_status_ocsp_resp(s, respder, ...)
"respder" is released from within the API, thus  line above : OPENSSL_free(respder); should be inside SSL_set_tlsext_status_ocsp_resp() function itself.
Is that correct?

Reply:
In original implementation "OPENSSL_free(respder);" line from above was not there.
Memory leak is because your new API for SSL_set_tlsext_status_ocsp_resp() have different data holder internally, and the data from "respder" is copied, but not freed.
You use same "sc->ext.ocsp.resp" internally, but for other purposes, i.e. not to hold "respder" to be deleted later by the openssl. 
So my proposal is to move this line "OPENSSL_free(respder);" inside your new implementation of SSL_set_tlsext_status_ocsp_resp(). In this way, you will preserve exact behavior of the old API.
Because if someone already use this function, and update to new API, it will have memory leak, if it does not notice that he need to do this additional free.

I'm currently changing my PR for multiple stapling responses for TLS 1.2 to use your new API SSL_set_tlsext_status_ocsp_resp_ex and your internal data holder, so merging our 2 PRs will be much more easy....if they are accepted at all :) Will commit next week.
Regards

Reply:
This is still unresolved?

Reply:
That's wrong! It is breaking the API contract because it will leak with current callers. With the existing implementation the caller cannot assume it can use the respder once it calls SSL_set_tlsext_status_ocsp_resp() and the call succeeds.